src/node_buffer.cc

```diff
@@ -675,6 +675,48 @@ void Fill(const FunctionCallbackInfo<Value>& args) {
   }
 }
 
+void StringWriteWithoutBuffer(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = args.GetIsolate();
+
+  if (!args[0]->IsString())
+    return env->ThrowTypeError("Source must be a string");
+  enum encoding source_encoding = ParseEncoding(isolate, args[1], UTF8);
+  enum encoding target_encoding = ParseEncoding(isolate, args[2], UTF8);
+
+  Local<String> str = args[0]->ToString(isolate);
+
+  if (source_encoding == HEX && str->Length() % 2 != 0)
+    return env->ThrowTypeError("Invalid hex string");
+
+  const size_t length = StringBytes::Size(isolate, str, source_encoding);
+  size_t actual = 0;
+  char* data = nullptr;
+
+  if (length > 0) {
+    data = static_cast<char*>(BUFFER_MALLOC(length));
+
+    if (data == nullptr)
+      return env->ThrowError("Unable to encode string");
+
+    actual = StringBytes::Write(isolate, data, length, str, source_encoding);
+    CHECK(actual <= length);
+
+    if (actual == 0) {
+      free(data);
+      data = nullptr;
+    } else if (actual < length) {
+      data = static_cast<char*>(realloc(data, actual));
+      CHECK_NE(data, nullptr);
+    }
+  }
+
+  args.GetReturnValue().Set(StringBytes::Encode(isolate,
+                                                data,
+                                                target_encoding));
+
+  free(data);
+}
```

Comment:
Possible future optimization: have a small (say 1,024 bytes) stack-allocated buffer to avoid the malloc/realloc/free dance for small inputs.